Sources/ClientRuntime/Networking/Streaming/StreamReader.swift

```diff
@@ -7,14 +7,31 @@ import AwsCommonRuntimeKit
 import class Foundation.FileHandle
 import class Foundation.FileManager
 
+/// Protocol for reading and writing to a stream.
 public protocol StreamReader: AnyObject {
```

Comment:
ooc, why the AnyObject restriction?

Reply:
if I have to guess, we don't want to copy data, and by mistake we could end up implementing it via a struct which is going to be a bad idea.

Backing/internal stream must always be a reference type.

Reply:
Gotcha, I think this is unnecessary and is an implementation detail leaking to the interface. It is quite possible to conform to this protocol without needing a class (a struct could always have a property that is a reference type where mutated values are stored). If this protocol is going to be very prescriptive about the implementation, then it seems like its too rigid to be a protocol and instead should be a concrete object that allows for configuration/customization. 

Related to Josh's comment earlier regarding StreamReader being the wrong abstraction, we should tackle this in follow-up work.